crates/boojum/src/gadgets/u512/mod.rs

```diff
@@ -216,6 +226,153 @@ impl<F: SmallField> UInt512<F> {
         (result, borrow_out)
     }
 
+    /// Multiplies a number by 2^{32}. Panics if the number overflows.
+    #[must_use]
+    pub fn must_mul_by_2_pow_32<CS: ConstraintSystem<F>>(&self, cs: &mut CS) -> Self {
+        let boolean_true = Boolean::allocated_constant(cs, true);
+        let last_limb_zero = self.inner[15].is_zero(cs);
+        Boolean::enforce_equal(cs, &last_limb_zero, &boolean_true);
+
+        let mut new_inner = self.inner;
+        new_inner.copy_within(0..15, 1);
+        new_inner[0] = UInt32::zero(cs);
+
+        Self { inner: new_inner }
+    }
+
+    /// Returns `true` if `self >= other`, and `false` otherwise.
+    /// Here, `self` and `other` are represented as `UInt512<F>` and `UInt256<F>` respectively.
+    pub fn geq_than_u256<CS>(&self, cs: &mut CS, other: &UInt256<F>) -> Boolean<F>
```

Comment:
I think here could be optimization for this; we have a special type of gate UIntXAddGate 
see u16_long_subtraction_noborrow_must_borrow

Reply:
I see that it is also used in u16 and u32 implementations. However, can't use it in u256 directly because max supported width for this gate is 32. It is though used under the hood in `UInt256::overflowing_sub` -> `UInt32::overflowing_sub_with_borrow_in`, so I suppose the optimization via this gate already holds 🙂